File: lab1_cpp/test/BoardTests.cpp
#include "doctest.hpp"
#include "Board.hpp"
#include <sstream>
#include <algorithm>

TEST_CASE("testConstructor")
{
	// Arrange
	std::string expected = 	" 23\n"
							"145\n"
							"678\n";

	// Act
	Board b("023145678");
	std::stringstream ss;
	ss << b;
	std::string bStr = ss.str();

	// Assert
	CHECK(bStr == expected);
}

TEST_CASE("testConstructor2")
{
	// Arrange
	std::string expected = 	"123\n"
							"485\n"
							"67 \n";

	// Act
	Board b("123485670");
	std::stringstream ss;
	ss << b;
	std::string bStr = ss.str();

	// Assert
	CHECK(bStr == expected);
}

TEST_CASE("getCost should return 1")
{
	// Arrange
	Board b0("123456780");
	Board b1("087654321");
	double expected = 1.0;

	// Act
	double cost0 = b0.getCost();
	double cost1 = b1.getCost();

	// Assert
	CHECK(cost0 == expected);
	CHECK(cost1 == expected);
}

TEST_CASE("should be equal")
{
	// Arrange
	Board b0("123456780");
	Board b1("123456780");
	Board b2("137245086");
	Board b3("137245086");

	// Act
	bool simmetry0 = (b0 == b1 && b1 == b0);
	bool simmetry1 = (b2 == b3 && b3 == b2);

	// Arrange
	CHECK(simmetry0);
	CHECK(simmetry1);
}

TEST_CASE("should not be equal")
{
	// Arrange
	Board b0("123456780");
	Board b1("137245086");
	Board b2("137245680");

	// Act
	bool simmetry0 = (b0 == b1 || b1 == b0);
	bool simmetry1 = (b0 == b2 || b2 == b0);
	bool simmetry2 = (b1 == b2 || b2 == b1);

	// Arrange
	CHECK(!simmetry0);
	CHECK(!simmetry1);
	CHECK(!simmetry2);
}

TEST_CASE("should have hash properties")
{
	// Arrange
	Board b0("123456780");
	Board b1("123456780");
	Board b2("137245086");
	Board b3("137245086");
	
	// Act
	bool simmetry0 = (b0 == b1 && b1 == b0);
	bool simmetry1 = (b2 == b3 && b3 == b2);
	std::size_t hash0 = b0.hash();
	std::size_t hash1 = b1.hash();
	std::size_t hash2 = b2.hash();
	std::size_t hash3 = b3.hash();

	// Arrange
	CHECK(simmetry0);
	CHECK(simmetry1);
	CHECK(hash0 == hash1);
	CHECK(hash2 == hash3);
	CHECK(hash0 != hash2);
}

TEST_CASE("should be goal when is equal")
{
	// Arrange
	Board b0("123456780");
	Board b1("123456780");
	Board b2("137245086");
	Board b3("137245086");
	
	// Act
	bool isGoal0 = b0.isGoal(b1);
	bool isGoal1 = b1.isGoal(b2);
	bool isGoal2 = b2.isGoal(b3);

	// Arrange
	CHECK(isGoal0);
	CHECK(!isGoal1);
	CHECK(isGoal2);
}

bool test_find(std::vector<std::unique_ptr<ILayout>>::iterator begin, std::vector<std::unique_ptr<ILayout>>::iterator end, const ILayout& expected)
{
	for (std::vector<std::unique_ptr<ILayout>>::iterator it = begin; it != end; it++)
		if (**it == expected)
			return true;
	return false;
}

TEST_CASE("should get 4 children")
{
	// Arrange
	Board b("134208657");
	Board expectedChild0("104238657");
	Board expectedChild1("134280657");
	Board expectedChild2("134028657");
	Board expectedChild3("134258607");

	// Act
	std::vector<std::unique_ptr<ILayout>> children = b.children();

	// Arrange
	CHECK(children.size() == 4);
	CHECK(test_find(children.begin(), children.end(), expectedChild0));
	CHECK(test_find(children.begin(), children.end(), expectedChild1));
	CHECK(test_find(children.begin(), children.end(), expectedChild2));
	CHECK(test_find(children.begin(), children.end(), expectedChild3));
}

TEST_CASE("should get 2 children on bottom left")
{
	// Arrange
	Board b("634218057");
	Board expectedChild0("634018257");
	Board expectedChild1("634218507");

	// Act
	std::vector<std::unique_ptr<ILayout>> children = b.children();

	// Arrange
	CHECK(children.size() == 2);
	CHECK(test_find(children.begin(), children.end(), expectedChild0));
	CHECK(test_find(children.begin(), children.end(), expectedChild1));
}

TEST_CASE("should get 2 children on top right")
{
	// Arrange
	Board b("630218357");
	Board expectedChild0("603218357");
	Board expectedChild1("638210357");

	// Act
	std::vector<std::unique_ptr<ILayout>> children = b.children();

	// Arrange
	CHECK(children.size() == 2);

	CHECK(test_find(children.begin(), children.end(), expectedChild0));
	CHECK(test_find(children.begin(), children.end(), expectedChild1));
}

File: lab2/src/test/BoardTests.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.List;

public class BoardTests
{
	@Test
	public void testConstructor()
	{
		Board b = new Board("023145678");
		StringWriter writer = new StringWriter();
		PrintWriter pw = new PrintWriter ( writer ) ;
		pw.println(" 23");
		pw.println("145");
		pw.println("678");
		assertEquals(writer.toString(), b.toString());
		pw.close();
	}

	@Test
	public void testConstructor2()
	{
		Board b = new Board("123485670");
		StringWriter writer = new StringWriter();
		PrintWriter pw = new PrintWriter (writer) ;
		pw.println("123");
		pw.println("485");
		pw.println("67 ");
		assertEquals(writer.toString(), b.toString());
		pw.close();
	}

	@Test
	public void getCostShoultReturnOne()
	{
		// Arrange
		Board b0 = new Board("123456780");
		Board b1 = new Board("087654321");
		double expected = 1.0d;

		// Act
		double cost0 = b0.getCost();
		double cost1 = b1.getCost();

		// Assert
		assertEquals(expected, cost0);
		assertEquals(expected, cost1);
	}

	@Test
	public void shouldEquals()
	{
		// Arrange
		Board b0 = new Board("123456780");
		Board b1 = new Board("123456780");
		Board b2 = new Board("137245086");
		Board b3 = new Board("137245086");
		
		// Act
		boolean simmetry0 = b0.equals(b1) && b1.equals(b0);
		boolean simmetry1 = b2.equals(b3) && b3.equals(b2);

		// Arrange
		assertTrue(simmetry0);
		assertTrue(simmetry1);
	}

	@Test
	public void shouldNotEquals()
	{
		// Arrange
		Board b0 = new Board("123456780");
		Board b1 = new Board("137245086");
		Board b2 = new Board("137245680");
		
		// Act
		boolean simmetry0 = b0.equals(b1) || b1.equals(b0);
		boolean simmetry1 = b0.equals(b2) || b2.equals(b0);
		boolean simmetry2 = b1.equals(b2) || b2.equals(b1);

		// Arrange
		assertFalse(simmetry0);
		assertFalse(simmetry1);
		assertFalse(simmetry2);
		assertNotEquals(b0, "123456780");
	}

	@Test
	public void shouldHaveHashProperties()
	{
		// Arrange
		Board b0 = new Board("123456780");
		Board b1 = new Board("123456780");
		Board b2 = new Board("137245086");
		Board b3 = new Board("137245086");
		
		// Act
		boolean simmetry0 = b0.equals(b1) && b1.equals(b0);
		boolean simmetry1 = b2.equals(b3) && b3.equals(b2);
		int hash0 = b0.hashCode();
		int hash1 = b1.hashCode();
		int hash2 = b2.hashCode();
		int hash3 = b3.hashCode();

		// Arrange
		assertTrue(simmetry0);
		assertTrue(simmetry1);
		assertEquals(hash0, hash1);
		assertEquals(hash2, hash3);
		assertNotEquals(hash0, hash2);
	}

	@Test
	public void shouldBeGoalWhenEquals()
	{
		// Arrange
		Board b0 = new Board("123456780");
		Board b1 = new Board("123456780");
		Board b2 = new Board("137245086");
		Board b3 = new Board("137245086");
		
		// Act
		boolean isGoal0 = b0.isGoal(b1);
		boolean isGoal1 = b1.isGoal(b2);
		boolean isGoal2 = b2.isGoal(b3);

		// Arrange
		assertTrue(isGoal0);
		assertFalse(isGoal1);
		assertTrue(isGoal2);
	}

	@Test
	public void shouldGetFourChildren()
	{
		// Arrange
		Board b = new Board("134208657");
		Board expectedChild0 = new Board("104238657");
		Board expectedChild1 = new Board("134280657");
		Board expectedChild2 = new Board("134028657");
		Board expectedChild3 = new Board("134258607");

		// Act
		List<ILayout> children = b.children();

		// Arrange
		assertEquals(children.size(), 4);
		assertTrue(children.contains(expectedChild0));
		assertTrue(children.contains(expectedChild1));
		assertTrue(children.contains(expectedChild2));
		assertTrue(children.contains(expectedChild3));
	}

	@Test
	public void shouldGetTwoChildrenOnBottomLeft()
	{
		// Arrange
		Board b = new Board("634218057");
		Board expectedChild0 = new Board("634018257");
		Board expectedChild1 = new Board("634218507");

		// Act
		List<ILayout> children = b.children();

		// Arrange
		assertEquals(children.size(), 2);
		assertTrue(children.contains(expectedChild0));
		assertTrue(children.contains(expectedChild1));
	}

	@Test
	public void shouldGetTwoChildrenOnTopRight()
	{
		// Arrange
		Board b = new Board("630218357");
		Board expectedChild0 = new Board("603218357");
		Board expectedChild1 = new Board("638210357");

		// Act
		List<ILayout> children = b.children();

		// Arrange
		assertEquals(children.size(), 2);
		assertTrue(children.contains(expectedChild0));
		assertTrue(children.contains(expectedChild1));
	}
}
